config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,100 @@ type BareMetalPlatformStatus struct {
 	NodeDNSIP string `json:"nodeDNSIP,omitempty"`
 }
 
+// ControlPlaneBGPPeer describes a BGP peer that the control plane should peer with.
+type ControlPlaneBGPPeer struct {
+	// asn is the Autonomous System Number of the peer.
+	//
+	// +optional
+	ASN string `json:"asn,omitempty"`
+
+	// ip is the IP address of the peer, as reachable from the Control
+	// plane machine. It may be either IPv4 or IPv6.
+	//
+	// +kubebuilder:validation:Required
+	// + ---
+	// + {IPv4 or IPv6} validation in crd.yaml-patch
+	IP string `json:"ip"`
+
+	// password for BGP authentication against the peer.
+	//
+	// +optional
+	Password string `json:"password,omitempty"`
+}
+
+// ControlPlaneBGPSpeaker describes the BGP speaker configuration that will
+// be used for the control plane.
+type ControlPlaneBGPSpeaker struct {
+	// subnetID is the subnet which this BGP configuration configuration
+	// applies to.
+	//
+	// +kubebuilder:validation:Required
+	SubnetID string `json:"subnetID"`
```

Comment:
I think this is too opinionated to be in the ControlPlaneBGPSpeaker and I think this should be up to each platform where the BGP configuration applies to (here a subnet, but for VMware? Baremetal?).

Reply:
@EmilienM I think you're suggesting this should be reusable across multiple providers, which sounds like a good goal. However, for OpenStack we are going to want this reference to be by SubnetID. Where does the SubnetID go?

Off the top of my head I can think of 2 general approaches:

1. There's already an idiom in OpenShift of cloud-specific discriminated unions. In this case the discriminator would be the platform type in Infrastructure, but we'd still end up with an entry here for every supported cloud provider, and we'd match based on whichever was in use.

2. We turn this into a generic reference, probably a name. We then need a mapping elsewhere of subnetID to name. Where would that be?